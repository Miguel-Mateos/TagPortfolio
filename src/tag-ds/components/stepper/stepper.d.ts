import React from 'react';
declare type TStatus = 'success' | 'pending' | 'disabled' | 'error' | 'default' | 'active';
export interface IStep {
    name: string;
    detail?: string;
    status?: TStatus;
}
interface IStepperProps {
    /**
     * List of steps with name, detail and status, being the status one of:
     * success, pending, disabled, error, default, active
     */
    steps: IStep[];
    /**
     * direction of the stepper
     */
    vertical?: boolean;
    numbered?: boolean;
    onClickStep?: (stepIndex: number) => void;
    className?: string;
    [others: string]: any;
}
export declare const Stepper: React.FC<IStepperProps>;
export default Stepper;
