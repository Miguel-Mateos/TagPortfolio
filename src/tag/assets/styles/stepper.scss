.tag-ds {
  .stepper,
  &.stepper {
    max-width: 100%;
    display: flex;
    align-items: center;
    gap: var(--unit);
    &.vertical {
      flex-direction: column;
    }
  }

  span.stepper-step-icon,
  .stepper-step-icon_numbered {
    @include font(medium);
    color: var(--neutral800);
    height: calc(3 * var(--unit));
    width: calc(3 * var(--unit));
    border-radius: 50%;
    border: calc(var(--unit) / 4) solid var(--neutral400);
    display: flex;
    justify-content: center;
    align-items: center;
  }

  .stepper-step-container {
    flex: 1 1 auto;
    border-top: calc(var(--unit) / 4) solid var(--neutral400);
    padding-top: calc(1.25 * var(--unit));
    display: flex;
    align-items: center;
    &.vertical {
      padding-top: 0;
      border-top: none;
    }
    &.last {
      align-items: flex-start;
      .stepper-step-icon-container_vertical {
        align-items: inherit;
        &::after {
          content: none;
        }
      }
    }

    &.active {
      border-color: var(--primary400);
      & .stepper-text-content {
        color: var(--neutral800);
      }
      & .stepper-step-icon,
      .stepper-step-icon_numbered {
        color: var(--neutral800);
        border-color: var(--primary400);
        border-radius: 50%;
      }
      & .stepper-step-icon-container_vertical {
        &::after {
          background-color: var(--primary400);
        }
      }
    }

    &.success {
      border-color: var(--success400);
      & .stepper-text-content {
        color: var(--neutral700);
        &:hover {
          color: var(--neutral800);
        }
      }
      & .stepper-step-icon,
      .stepper-step-icon_numbered {
        color: var(--neutral0);
        border-color: transparent;
        background-color: var(--success400);
        border-radius: 50%;
      }

      & .stepper-step-icon_numbered:hover {
        &::after {
          height: calc(3 * var(--unit));
          width: calc(3 * var(--unit));
          border-color: transparent;
          background-color: var(--success400);
          border-radius: 50%;
          color: var(--neutral0);
          font-family: 'Material Icons';
          position: absolute;
          content: '\e876';
          display: flex;
          justify-content: center;
          align-items: center;
        }
      }
      & .stepper-step-icon-container_vertical {
        &::after {
          background-color: var(--success400);
        }
      }
    }
    &.pending {
      border-color: var(--neutral400);
      & .stepper-step-icon,
      .stepper-step-icon_numbered {
        color: var(--neutral800);
        border-color: var(--neutral400);
        border-radius: 50%;
      }
    }
    &.disabled {
      border-color: var(--neutral200);
      & .stepper-step-icon,
      .stepper-step-icon_numbered {
        border-color: var(--neutral200);
        border-radius: 50%;
      }

      & .stepper-title,
      & .stepper-subtitle {
        color: var(--neutral500);
      }
    }
    &.error {
      border-color: var(--error400);
      & .stepper-text-content {
        color: var(--neutral700);
        &:hover {
          color: var(--neutral800);
        }
      }
      & .stepper-step-icon,
      .stepper-step-icon_numbered {
        border-color: transparent;
        color: var(--neutral0);
        background-color: var(--error400);
        border-radius: 50%;
      }
      & .stepper-step-icon-container_vertical {
        &::after {
          background-color: var(--error400);
        }
      }
    }
  }

  .stepper-title {
    margin: 0;
    @include font(base);
  }

  .stepper-subtitle {
    margin: 0;
    font-size: calc(1.5 * var(--unit));
  }

  .stepper-step-icon-container_vertical {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    &::after {
      margin: var(--unit) 0;
      content: '';
      height: calc(6 * var(--unit));
      width: calc(var(--unit) / 4);
      background-color: var(--neutral200);
    }
  }

  .stepper-text-content {
    margin-left: var(--unit);
    color: var(--neutral700);
  }

  .stepper-text-content {
    align-self: flex-start;
  }

  .stepper-mobile {
    padding: calc(2 * var(--unit));
    width: fit-content;
  }

  .stepper-mobile-text-container {
    display: flex;
    align-items: center;
    justify-content: space-between;
  }

  .stepper-mobile-steps {
    @include font(small);
  }

  .stepper-mobile-progress {
    margin-top: calc(var(--unit) / 2);
    width: calc(41 * var(--unit));
    height: var(--unit);
    background-color: var(--neutral200);
    border-radius: var(--unit);
    position: relative;
    overflow: hidden;
  }
  .stepper-mobile-progress_fill {
    position: absolute;
    width: calc(41 * var(--unit));
    height: var(--unit);
    background-color: var(--accent500);
    border-radius: var(--unit);
  }
}

