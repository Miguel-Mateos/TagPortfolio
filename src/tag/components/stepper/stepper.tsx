import React, {useEffect, useState} from 'react';

type TStatus = 'success' | 'pending' | 'disabled' | 'error' | 'default' | 'active';
export interface IStep {
  name: string;
  detail?: string;
  status?: TStatus;
}

interface IStepperProps {
  /**
   * List of steps with name, detail and status, being the status one of:
   * success, pending, disabled, error, default, active
   */
  steps: IStep[];
  /**
   * direction of the stepper
   */
  vertical?: boolean;
  numbered?: boolean;
  onClickStep?: (stepIndex: number) => void;
  className?: string;
  [others: string]: any;
}

export const Stepper: React.FC<IStepperProps> = ({steps, vertical, numbered, onClickStep, className, ...rest}) => {
  const [mobile, setMobile] = useState<boolean>(window.innerWidth < 672);
  useEffect(() => {
    window.addEventListener('resize', () => {
      if (window.innerWidth < 672) setMobile(true);
      else setMobile(false);
    });
    return () => window.removeEventListener('resize', () => null);
  }, []);

  const handlerIcon = (str: TStatus = 'default') => {
    if (str === 'success') return 'done';
    return str;
  };

  const onClickStepItem = (index: number) => {
    if (typeof onClickStep === 'function') onClickStep(index);
  };

  if (mobile) {
    const activeStep = steps.findIndex((step) => step.status === 'active');
    return (
      <div className="stepper-mobile">
        <div className="stepper-mobile-text-container">
          <p className="stepper-title">{steps[activeStep]?.name}</p>
          <div className="stepper-mobile-steps">
            {activeStep}/{steps.length}
          </div>
        </div>
        <div className="stepper-mobile-progress">
          <div
            className="stepper-mobile-progress_fill"
            style={{transform: `translateX(-${100 - (activeStep / steps.length) * 100}%)`}}
          />
        </div>
      </div>
    );
  }
  if (vertical) {
    return (
      <div
        className={`stepper_vertical ${className || ''}`}
        data-testid={rest && rest['data-testid'] ? rest['data-testid'] : undefined}
      >
        {steps.map((step, idx) => (
          <div
            className={`stepper-step-container vertical ${step.status}  ${idx === steps.length - 1 ? 'last' : ''}`}
            key={idx}
            onClick={() => onClickStepItem(idx)}
          >
            <div className="stepper-step-icon-container_vertical">
              {numbered ? (
                <span className="stepper-step-icon_numbered">{idx}</span>
              ) : (
                <span className="material-icons stepper-step-icon">{handlerIcon(step.status)}</span>
              )}
            </div>
            <div className="stepper-text-content">
              <p className="stepper-title">{step.name}</p>
              <p className="stepper-subtitle">{step.detail}</p>
            </div>
          </div>
        ))}
      </div>
    );
  }

  return (
    <div
      className={`stepper ${className || ''}`}
      data-testid={rest && rest['data-testid'] ? rest['data-testid'] : undefined}
    >
      {steps.map((step: IStep, idx: number) => (
        <div
          className={`stepper-step-container ${step.status}`}
          key={idx}
          onClick={() => onClickStepItem(idx)}
          data-testid={rest && rest['data-testid'] ? `${rest['data-testid']}-item-${idx}` : undefined}
        >
          <div className="stepper-step-icon-container">
            {numbered ? (
              <span className="stepper-step-icon_numbered">{idx}</span>
            ) : (
              <span className="material-icons stepper-step-icon">{handlerIcon(step.status)}</span>
            )}
          </div>
          <div className="stepper-text-content">
            <p className="stepper-title">{step.name}</p>
            <p className="stepper-subtitle">{step.detail}</p>
          </div>
        </div>
      ))}
    </div>
  );
};

export default Stepper;

