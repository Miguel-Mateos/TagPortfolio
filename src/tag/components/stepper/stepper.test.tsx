import React from 'react';
import { IStep, Stepper } from './stepper';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';

const steps: IStep[] = [
  {
    name: 'Step Name',
    detail: 'optional',
    status: 'success'
  },
  {
    name: 'Step Name',
    detail: 'optional',
    status: 'pending'
  },
  {
    name: 'Step Name',
    detail: 'optional',
    status: 'error'
  },
  {
    name: 'Step Name',
    detail: 'optional',
    status: 'disabled'
  },
  {
    name: 'Step Name',
    detail: 'optional',
    status: 'active'
  }
];

const StepperTest = (props: any) => (
  <Stepper steps={steps} direction={props.direction}/>
);

test('Render Stepper and match snapshot', () => {
  const component = renderer.create(<StepperTest />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('render stepper vertical', () => {
  const component = render(<StepperTest direction='vertical' />);
  expect(component).toBeDefined();
});

test('render with mobile', () => {
  // eslint-disable-next-line no-global-assign
  window = Object.assign(window, { innerWidth: 600 });
  const component = render(<StepperTest />);
  expect(component).toBeDefined();
});