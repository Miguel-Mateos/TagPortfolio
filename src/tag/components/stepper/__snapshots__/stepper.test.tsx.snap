// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Stepper and match snapshot 1`] = `
<div
  className="stepper "
>
  <div
    className="stepper-step-container success"
    onClick={[Function]}
  >
    <div
      className="stepper-step-icon-container"
    >
      <span
        className="material-icons stepper-step-icon"
      >
        done
      </span>
    </div>
    <div
      className="stepper-text-content"
    >
      <p
        className="stepper-title"
      >
        Step Name
      </p>
      <p
        className="stepper-subtitle"
      >
        optional
      </p>
    </div>
  </div>
  <div
    className="stepper-step-container pending"
    onClick={[Function]}
  >
    <div
      className="stepper-step-icon-container"
    >
      <span
        className="material-icons stepper-step-icon"
      >
        pending
      </span>
    </div>
    <div
      className="stepper-text-content"
    >
      <p
        className="stepper-title"
      >
        Step Name
      </p>
      <p
        className="stepper-subtitle"
      >
        optional
      </p>
    </div>
  </div>
  <div
    className="stepper-step-container error"
    onClick={[Function]}
  >
    <div
      className="stepper-step-icon-container"
    >
      <span
        className="material-icons stepper-step-icon"
      >
        error
      </span>
    </div>
    <div
      className="stepper-text-content"
    >
      <p
        className="stepper-title"
      >
        Step Name
      </p>
      <p
        className="stepper-subtitle"
      >
        optional
      </p>
    </div>
  </div>
  <div
    className="stepper-step-container disabled"
    onClick={[Function]}
  >
    <div
      className="stepper-step-icon-container"
    >
      <span
        className="material-icons stepper-step-icon"
      >
        disabled
      </span>
    </div>
    <div
      className="stepper-text-content"
    >
      <p
        className="stepper-title"
      >
        Step Name
      </p>
      <p
        className="stepper-subtitle"
      >
        optional
      </p>
    </div>
  </div>
  <div
    className="stepper-step-container active"
    onClick={[Function]}
  >
    <div
      className="stepper-step-icon-container"
    >
      <span
        className="material-icons stepper-step-icon"
      >
        active
      </span>
    </div>
    <div
      className="stepper-text-content"
    >
      <p
        className="stepper-title"
      >
        Step Name
      </p>
      <p
        className="stepper-subtitle"
      >
        optional
      </p>
    </div>
  </div>
</div>
`;
